pkg/dataobj/internal/encoding/decoder_range.go

```diff
@@ -0,0 +1,275 @@
+package encoding
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/dataset"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/filemd"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/logsmd"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/streamsmd"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/result"
+)
+
+// rangeReader is an interface that can read a range of bytes from an object.
+type rangeReader interface {
+	// Size returns the full size of the object.
+	Size(ctx context.Context) (int64, error)
+
+	// ReadRange returns a reader over a range of bytes. Callers may create
+	// multiple current instance of ReadRange.
+	ReadRange(ctx context.Context, offset int64, length int64) (io.ReadCloser, error)
+}
+
+type rangeDecoder struct {
```

Comment:
This is more or less a copy/paste from the old bucket decoder but it uses rangeReader instead.